### PR TITLE
Timespinner:  New options from TS Rando v1.25 + Logic fix

### PR DESCRIPTION
**Pull request by Jarno458**
## What is this fixing or adding?
A logic fix for Sealed Caves

2 new options
- "EnemyRando": randomize enemies in rooms
- "PresentAccessWithWheelAndSpindle": allow using wheel + spindle in inverted seeds to warp to present in refugee camp

2 new floodable areas:
- Lab
- Lake Serene Bridge

## How was this tested?
Ran 100 generations locally, checked a few spoiler logs
Is Available as .apworld and has been run by a few community members

Included #2373 & #2374

### CODE REVIEW

**Comment by black-sliver** [comment]
I have 2 comments:

* it would be nice to get logic fixes and feature additions in separate PRs. Not sure if we want to merge this into the upcoming release as it is.

* not a blocker, but there is potential performance gain in short circuiting the conditions in the lambdas:
```python
lambda state: state.has('Water Mask', player) or not flooded.flood_lab)
```
would most likely be faster as
```python
lambda state: not flooded.flood_lab or state.has('Water Mask', player))
```
because the flood lookup is always faster and more likely to be false (because it'd be false for default settings AND for 67% of the default random flooding)

**Comment by Jarno458** [comment]
I agree on swapping those parameters, I did that also in the past. Timespinner is already rather fast on generation 

However setting the rule based on the flag, while fasterm it likely wont make the code any more readable and be a very high risk change as it should be done for all those logic rules. I don feel comfortable to make that change atm for the very small performance gain it will likely give

**Comment by Jarno458** [comment]
Iv updated the parameter order, as for when to merge this. not up to me, but this PR was already created before 0.4.2 was live, its not a bugfix for something that added in 0.4.2 its a bugfix for things added in 0.4.1

**Comment by Jarno458** [comment]
Included #2373 & #2374

File: worlds/timespinner/Regions.py [hunk 1]
```diff
@@ -213,6 +213,17 @@ def create_location(player: int, location_data: LocationData, region: Region) ->
     return location
 
 
+def create_regions(world: MultiWorld, player: int, locations_per_region: Dict[str, List[LocationData]],
+                    region_names: List[str]) -> Dict[str, Region]:
+    
+    regions: Dict[str, Region] = {}
+
+    for name in region_names:
+        regions[name] = create_region(world, player, locations_per_region, name)
```
Comment (black-sliver) [inline worlds/timespinner/Regions.py:222]: `locations_per_region` seems to be unused?

File: worlds/timespinner/Regions.py [hunk 2]
```diff
@@ -204,7 +202,9 @@ def throwIfAnyLocationIsNotAssignedToARegion(regions: List[Region], regionNames:
 
 def create_location(player: int, location_data: LocationData, region: Region) -> Location:
     location = Location(player, location_data.name, location_data.code, region)
-    location.access_rule = location_data.rule
+
+    if (location_data.rule):
```
Comment (black-sliver) [inline worlds/timespinner/Regions.py:206]:
```suggestion
    if location_data.rule:
```

**Review by black-sliver — COMMENTED** [review]
Looks good other than the two small things below.

How much was this tested?

File: worlds/timespinner/__init__.py [hunk 3]
```diff
@@ -39,21 +39,19 @@ class TimespinnerWorld(World):
     option_definitions = timespinner_options
     game = "Timespinner"
     topology_present = True
-    data_version = 11
+    data_version = 12
```
Comment (alwaysintreble) [inline worlds/timespinner/__init__.py:42]: would probably be better to just remove this?

**Review by alwaysintreble — APPROVED** [review]
Could probably use some optimization but I know you're already working on that and moving to the new options API so lgtm

File: worlds/timespinner/__init__.py (hunk 3 above)
Comment (Jarno458) [reply worlds/timespinner/__init__.py:42]: Not sure if we already can remove that?

**Comment by Jarno458** [comment]
> Looks good other than the two small things below.
> 
> How much was this tested?

Well the new options have been available as an .apworld for a couple on months now. didnt get complains about that, but hard to tell how much people actually used it

As for the generation and optimizations i tested those locally be generating atleast 50 times using a random yaml to hit different combinations of options

File: worlds/timespinner/Regions.py (hunk 1 above)
Comment (Jarno458) [reply worlds/timespinner/Regions.py:222]: its used, its passed on into the create_region method that will create the locations that belong to that region inside it

**Comment by Jarno458** [comment]
Drafted becase a potentional logic error was found

**Comment by Jarno458** [comment]
Fixed small logic bug, put back in ready for review

**Review by ThePhar — APPROVED** [review]

**Pull request merged**